projectq/setups/decompositions/cnot2rxx_test.py

```diff
@@ -0,0 +1,139 @@
+#   Copyright 2018 ProjectQ-Framework (www.projectq.ch)
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+"Tests for projectq.setups.decompositions.cnot2rxx.py."
+
+import pytest
+import numpy as np
+
+import projectq
+from projectq import MainEngine
+from projectq.backends import Simulator
+from projectq.cengines import (AutoReplacer, DecompositionRuleSet, DummyEngine,
+                               InstructionFilter)
+from projectq.meta import Control
+from projectq.ops import All, CNOT, CZ, Measure, X, Z
+
+from projectq.setups.decompositions import cnot2rxx
+
+def test_recognize_correct_gates():
+    """Test that recognize_cnot recognizes cnot gates. """
+    saving_backend = DummyEngine(save_commands=True)
+    eng = MainEngine(backend=saving_backend)
+    qubit1 = eng.allocate_qubit()
+    qubit2 = eng.allocate_qubit()
+    qubit3 = eng.allocate_qubit()
+    eng.flush()
+    # Create a control function in 3 different ways
+    CZ | (qubit1, qubit2)
+    with Control(eng, qubit2):
+        Z | qubit1
+        X | qubit1
+    with Control(eng, qubit2 + qubit3):
+        Z | qubit1
+    eng.flush()
+    eng.flush(deallocate_qubits=True)
+    for cmd in saving_backend.received_commands[4:7]:
+        assert cnot2rxx._recognize_cnot2(cmd)
+    for cmd in saving_backend.received_commands[7:9]:
+        assert not cnot2rxx._recognize_cnot2(cmd)
+
+def _decomp_gates(eng, cmd):
+    """ Test that the cmd.gate is a gate of class X """
+    g = cmd.gate
+    if len(cmd.control_qubits) == 1 and isinstance(cmd.gate, X.__class__):
+        return False
+    return True
+
+# ------------test_decomposition function-------------#
+# Creates two engines, correct_eng and test_eng. 
+# correct_eng implements CNOT gate. 
+# test_eng implements the decomposition of the CNOT gate.
+# correct_qb and test_qb represent results of these two
+# engines, respectively. 
+# The decomposition in this case only produces the 
+# same state as CNOT up to a global phase. 
+# test_vector and correct_vector represent the final 
+# wave states of correct_qb and test_qb. 
+# The dot product of correct_vector and test_vector
+# should have absolute value 1, if the two vectors are the
+# same up to a global phase.
+
+def test_decomposition():
```

Comment:
This test only tests the *_M decomposition rule. Please modify it so that it also covers the *_P decomposition rule.

To avoid duplicating code, you could parametrise this test function with a single parameter which would take one of 2 values stored in `cnot2rxx.all_defined_decomposition_rules`

Something like:
```python
@pytest.mark.parametrize("rule", cnot2rxx.all_defined_decomposition_rules)
def test_decomposition(rule):
    ...
    rule_set = DecompositionRuleSet(rules=[rule])
    ...
```